programs/bpf_loader/src/serialization.rs

```diff
@@ -99,59 +120,55 @@ impl Serializer {
                 self.fill_write(MAX_PERMITTED_DATA_INCREASE + BPF_ALIGN_OF_U128, 0)
                     .map_err(|_| InstructionError::InvalidArgument)?;
                 self.region_start += BPF_ALIGN_OF_U128.saturating_sub(align_offset);
+                // put the realloc padding in its own region
+                self.push_region(account.can_data_be_changed().is_ok());
             }
         }
 
-        Ok(())
+        Ok(vm_data_addr)
     }
 
-    fn push_account_region(
+    fn push_account_data_region(
         &mut self,
         account: &mut BorrowedAccount<'_>,
     ) -> Result<(), InstructionError> {
```

Comment:
how about adding this (yes, intentionally outside `if !account.get_data().is_empty()`, imho)?

```rust
debug_assert_eq!(self.region_start, self.buffer.len());
```